react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+
+~~~jsx
+const value = useMemo(
+  () => ({ someState, someFunction }),
+  [someState, someFunction]
+);
+
+return <Context.Provider value={value}>{children}</Context.Provider>;
+~~~
+
+### The useCallback hook
+
+The `useCallback` hook provides another way to memoize a value, not just **any** value like `useMemo`. It can only memoize a function. Did you see the previous snippet that we have with memoizing a function reference with `useMemo`?
+
+~~~jsx
+const handleClick = useMemo(
+  () => () => setCount((prevState) => prevState + 1),
+  []
+);
+// or
+const memoizedHandleClick = useMemo(() => handleClick, []);
+~~~
+
+With `useCallback`, we don't need to do that. It's specifically made for functions:
+
+~~~jsx
+import { useCallback } from "react";
+
+// Inside a component
+// Without useCallback
+const handleClick = () => setCount((prevState) => prevState + 1);
+// With useCallback
+const handleClick = useCallback(
+  () => setCount((prevState) => prevState + 1),
+  []
+);
+// or
+const memoizedHandleClick = useCallback(handleClick, []);
+~~~
+
+<a name="usememo-or-usecallback"></a>
+Yay, there's only one arrow function, and it's simpler to read. There's nothing extra to `useCallback` other than it only memoizes functions. So the main difference between `useMemo` and `useCallback` is just the type of value it returns.
+
+Which one should I use, then? Use `useMemo` for *any* value types, and use `useCallback` specifically for functions. At the end of the day, they both do similar things with a tiny difference, so use whatever you prefer.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. The article [When to useMemo and useCallback](https://kentcdodds.com/blog/usememo-and-usecallback) by Kent C. Dodds further introduces more examples of when to use `useMemo` and `useCallback` and why you shouldn't bother using it.
+2. We've only learned about a basic implementation of the `useRef` hook. For more examples about its usage and why we should be wary of using the hook (more on the links they provided in the guide), check out the [interactive guide of the React documentation](https://react.dev/reference/react/useRef) for `useRef`. Try your best to complete all the challenges!
+3. The article [useRef instead of querySelector in React](https://meje.dev/blog/useref-not-queryselector) by Caleb Olojo briefly tells some unexpected behaviors when trying to manipulate the DOM directly with DOM manipulation methods and why we should prefer `useRef` over other DOM manipulation methods like `querySelector`. Check it out!
+
+</div>
+
+### Knowledge check
+
+This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.
+
+- [Why should you prefer useRef hook over other DOM manipulation methods like querySelector?](https://meje.dev/blog/useref-not-queryselector)
```

Comment:
This article is already mentioned under the assignment point 3.

Reply:
This is more for knowledge check since this question is related to the article I've linked above

Reply:
My bad, again - mistook it for addl. resources.